server.js

```diff
@@ -0,0 +1,219 @@
+ *
+ * @param {import('vscode-languageserver').TextDocument} document
+ * @returns {import('vfile').VFile}
+ */
+function lspDocumentToVfile(document) {
+  return new VFile({
+    // VFile expects a file path or file URL object, but LSP provides a file URI
+    // as a string.
+    path: new URL(document.uri),
+    value: document.getText()
+  })
+}
+
+/**
+ * @param {import('vscode-languageserver').Connection} connection
+ * @param {import('vscode-languageserver').TextDocuments<import('vscode-languageserver-textdocument').TextDocument>} documents
+ * @param {import('./index.js').Options} options
+ */
+export function configureUnifiedLanguageServer(
+  connection,
+  documents,
+  {ignoreName, packageField, pluginPrefix, plugins, rcName, defaultSource}
+) {
+  /**
+   * Process various LSP text documents using unified and send back the
+   * resulting messages as diagnostics.
+   *
+   * @param {import('vscode-languageserver').TextDocument[]} textDocuments
+   * @param {boolean} alwaysStringify
+   * @returns {Promise<VFile[]>}
+   */
+  function processDocuments(textDocuments, alwaysStringify = false) {
+    return new Promise((resolve, reject) => {
+      engine(
+        {
+          alwaysStringify,
+          files: textDocuments.map((document) => lspDocumentToVfile(document)),
+          ignoreName,
+          packageField,
+          pluginPrefix,
+          plugins,
+          processor: unified(),
+          quiet: false,
+          rcName,
+          silentlyIgnore: true,
+          streamError: new PassThrough(),
+          streamOut: new PassThrough()
+        },
+        (error, code, context) => {
+          /* c8 ignore start */
+          if (error) {
+            reject(error)
+          } else {
+            resolve((context && context.files) || [])
+          }
+          /* c8 ignore end */
+        }
+      )
+    })
+  }
+
+  /**
+   * Process various LSP text documents using unified and send back the
+   * resulting messages as diagnostics.
+   *
+   * @param {import('vscode-languageserver').TextDocument[]} textDocuments
+   */
+  async function checkDocuments(...textDocuments) {
+    const documentVersions = new Map(
+      textDocuments.map((document) => [document.uri, document.version])
+    )
+    const files = await processDocuments(textDocuments)
+
+    for (const file of files) {
+      // VFile uses a file path, but LSP expects a file URL as a string.
+      const uri = String(pathToFileURL(file.path))
+      connection.sendDiagnostics({
+        uri,
+        version: documentVersions.get(uri),
+        diagnostics: file.messages.map((message) =>
+          vfileMessageToDiagnostic(message, defaultSource)
+        )
+      })
+    }
+  }
+
+  connection.onInitialize(() => ({
+    capabilities: {
+      textDocumentSync: TextDocumentSyncKind.Full,
+      documentFormattingProvider: true
+    }
+  }))
+
+  connection.onDocumentFormatting(async ({textDocument: {uri}}) => {
```

Comment:
I find such patterns hard to read and hard to debug (especially due to the double nesting). Is this some sort of event? I’d prefer intermediate names, e.g.:

```diff
-async ({textDocument: {uri}}) => {
+async (event) => {
+   const uri = event.textDocument.uri
```

Or:


```diff
-async ({textDocument: {uri}}) => {
+async (event) => {
+   const {uri} = event.textDocument
```